src/coreclr/jit/gentree.cpp

```diff
@@ -12685,63 +12685,132 @@ GenTree* Compiler::gtFoldExpr(GenTree* tree)
     if (opts.OptimizationDisabled())
     {
         return call;
     }
 
-    // Fetch id of the intrinsic.
-    const CorInfoIntrinsics methodID = info.compCompHnd->getIntrinsicID(call->gtCallMethHnd);
-
-    switch (methodID)
+    // Is this an intrinsic we can fold?
+    if (call->IsSpecialIntrinsic())
     {
-        case CORINFO_INTRINSIC_TypeEQ:
-        case CORINFO_INTRINSIC_TypeNEQ:
+        // Fetch id of the intrinsic.
+        const CorInfoIntrinsics methodID = info.compCompHnd->getIntrinsicID(call->gtCallMethHnd);
+        switch (methodID)
         {
-            noway_assert(call->TypeGet() == TYP_INT);
-            GenTree* op1 = call->gtCallArgs->GetNode();
-            GenTree* op2 = call->gtCallArgs->GetNext()->GetNode();
+            case CORINFO_INTRINSIC_TypeEQ:
+            case CORINFO_INTRINSIC_TypeNEQ:
+            {
+                noway_assert(call->TypeGet() == TYP_INT);
+                GenTree* op1 = call->gtCallArgs->GetNode();
+                GenTree* op2 = call->gtCallArgs->GetNext()->GetNode();
+
+                // If either operand is known to be a RuntimeType, this can be folded
+                GenTree* result = gtFoldTypeEqualityCall(methodID, op1, op2);
+                if (result != nullptr)
+                {
+                    return result;
+                }
+                break;
+            }
+
+            default:
+                break;
+        }
+
+        // Check for a new-style jit intrinsic.
+        const NamedIntrinsic ni = lookupNamedIntrinsic(call->gtCallMethHnd);
+
+        if (ni == NI_System_Enum_HasFlag)
+        {
+            GenTree* thisOp = call->gtCallThisArg->GetNode();
+            GenTree* flagOp = call->gtCallArgs->GetNode();
+            GenTree* result = gtOptimizeEnumHasFlag(thisOp, flagOp);
 
-            // If either operand is known to be a RuntimeType, this can be folded
-            GenTree* result = gtFoldTypeEqualityCall(methodID, op1, op2);
             if (result != nullptr)
             {
                 return result;
             }
-            break;
         }
+    }
 
-        default:
+    // There was no intrinsic to fold, try the helper path.
+    // At present, we only fold the helpers for casts.
+    const CorInfoHelpFunc helper       = eeGetHelperNum(call->gtCallMethHnd);
+    bool                  overflowCast = false;
+    bool                  fromUnsigned = false;
+    var_types             castToType   = call->TypeGet();
+
+    assert(castToType == genActualType(castToType));
+
+    switch (helper)
+    {
+        case CORINFO_HELP_ULNG2DBL:
+            fromUnsigned = true;
             break;
+        case CORINFO_HELP_LNG2DBL:
+        case CORINFO_HELP_DBL2INT:
+        case CORINFO_HELP_DBL2LNG:
+            break;
+        case CORINFO_HELP_DBL2UINT:
+            castToType = TYP_UINT;
+            break;
+        case CORINFO_HELP_DBL2ULNG:
+            castToType = TYP_ULONG;
+            break;
```

Comment:
I believe that you will only see these JIT helpers on the 32-bit targets (x86 and arm)
For the 64-bit targets we use a GT_CAST node.

Reply:
My understanding is the overflow helpers are used on all platforms (for the double -> integral conversions), and the `DBL2ULNG` is used on `x64` as well, so I didn't want to use `ifdef`s here as they would be duplicating the non-trivial logic in `fgMorphCast`. Do you think it would be worth doing that?

Reply:
I am OK with not adding #ifdef's for this